azure/Guidelines.md

```diff
@@ -0,0 +1,861 @@
+
+> :white_check_mark: **DO** Add RBAC roles for every service operation that requires accessing Storage scoped to the exact permissions.
+> 
+> :white_check_mark: **DO** Ensure that RBAC roles are backward compatible, and specifically, do not take away permissions from a role that would break the operation of the service. Any change of RBAC roles that results in a change of the service behavior is considered a breaking change.
+
+##### Handling 'downstream' errors
+It is not uncommon to rely on other services, e.g. storage, when implementing your service. Inevitably, the services you depend on will fail. In these situations, you can include the downstream error code and text in the inner-error of the response body. This provides a consistent pattern for handling errors in the services you depend upon.
+
+> :white_check_mark: **DO** include error from downstream services as the 'inner-error' section of the response body. 
+
+#### Working with files
+Generally speaking, there are two patterns that you will encounter when working with files; single file access, and file collections.
+
+##### Single file access
+Desiging an API for accessing a single file, depending on your scenario, is relatively straight forward.
+
+> :heavy_check_mark: **YOU MAY** use a Shared Access Signature [SAS](https://docs.microsoft.com/en-us/azure/storage/common/storage-sas-overview) to provide access to a single file. SAS is considered the minimum security for files and can be used in lieu of, or in addition to, RBAC.   
+
+> :ballot_box_with_check: **YOU SHOULD** if using HTTP (not HTTPS) document to users that all information is sent over the wire in clear text.  
+> 
+> :ballot_box_with_check: **YOU SHOULD** support managed identity using Azure Storage by default (if using Azure services). 
+
+###### File Versioning
+Depending on your requirements, there are scenarios where users of your service will require a specific version of a file. For example, you may need to keep track of configuration changes over time to be able to rollback to a previous state. In these scenarios, you will need to provide a mechanism for accessing a specific version.
+
+> :white_check_mark: **DO** Enable the customer to provide an ETag to specify a specific version of a file. 
+##### File Collections
+When your users need to work with multiple files, for example a document translation service, it will be important to provide them access to the collection, and its contents, in a consistent manner. Because there is no industry standard for working with with containers, these guidelines will recommend that you leverage Azure Storage. Following the guidelines above, you also want to ensure that you don't expose file system constructs, e.g. folders, and instead use storage constructs, e.g. blob prefixes.
```

Comment:
```suggestion
When your users need to work with multiple files, for example a document translation service, it will be important to provide them access to the collection, and its contents, in a consistent manner. Because there is no industry standard for working with containers, these guidelines will recommend that you leverage Azure Storage. Following the guidelines above, you also want to ensure that you don't expose file system constructs, e.g. folders, and instead use storage constructs, e.g. blob prefixes.
```

Reply:
Good catch. Thank you.